### PR TITLE
Research into thoop and squash 'refresh rate' protection [Victor Fernandez (City Game), Peter Ferrie, David Haywood]

### PR DESCRIPTION
**Pull request by DavidHaywood**
Research into thoop and squash 'refresh rate' protection [Victor Fernandez (City Game), Peter Ferrie, David Haywood]
 - prevents freeze when dying in stage 4 of Thunder Hoop, and bad text on continue in Squash
 
Made gaelcrpt a device rather than static (cleaner code for thoop, wrally, squash) [David Haywood]
 
Save gaelco sound registers properly (prevents bad loops etc. on state load in touchgo) [David Haywood]

Better handling of ds5002fp RNR for touchgo (previously no random numbers were returned) [David Haywood]

### CODE REVIEW

**Comment by DavidHaywood** [comment]
I've created a new PR and left https://github.com/mamedev/mame/pull/8903 for reference.  This also reverts the demotion of the game as it now works as expected.

File: src/mame/drivers/gaelco.cpp [hunk 1]
```diff
@@ -14,21 +14,77 @@ Year   Game                PCB            NOTES
...
+	  0xc368 sets 0xffe08e to 0x27.  Again the failure is
+	  not obvious at this point.
+
+	  0xffe08e is checked during player respawn after
+	  losing a life or continuing at 0x16d00, with an
+	  explicit compare against 0x27, if this condition is
+	  met, then the game will intentionally corrupt memory
+	  and crash.
+
+	  Many of these checks are done with obfusicated code
```
Comment (einstein95) [inline src/mame/drivers/gaelco.cpp:57]: * obfuscated

File: src/mame/drivers/gaelco.cpp [hunk 2]
```diff
@@ -14,21 +14,77 @@ Year   Game                PCB            NOTES
...
+	  after a continue.
+
+	  A refresh rate of 57.42, while not yet accurately
+	  measured, allows a video of thoop to stay in sync with
+	  MAME over a 10 minute period.
+
+	  No checks have been observed in Biomechanical Toy,
+	  the Maniac Square prototype, or the Last KM prototype.
+
+	  Big Karnk runs on a different board type and does fail
```
Comment (einstein95) [inline src/mame/drivers/gaelco.cpp:84]: \* Big Karnak

File: src/devices/cpu/mcs51/mcs51.cpp [hunk 3]
```diff
@@ -2059,8 +2059,13 @@ void mcs51_cpu_device::execute_run()
 
 		/* decrement the timed access window */
 		if (m_features & FEATURE_DS5002FP)
+		{
 			m_ds5002fp.ta_window = (m_ds5002fp.ta_window ? (m_ds5002fp.ta_window - 1) : 0x00);
 
+			if (m_ds5002fp.rnr_delay > 0)
+				m_ds5002fp.rnr_delay--;
```
Comment (happppp) [inline src/devices/cpu/mcs51/mcs51.cpp:2066]: I can understand that this random number generator is busy for a while, but it does not make sense that it has 1 delay per opcode. I don't know the opcode timings, but hypothetically let's say you have 160 opcodes each 10 cycles, meaning the random number generator is ready after 1600 cycles. Or you have 160 opcodes each 30 cycles, and it's ready after 4800 cycles?

In short, subtract m_inst_cycles for this delay thing?

Comment (DavidHaywood) [reply src/devices/cpu/mcs51/mcs51.cpp]: made the changes for the part you asked.  honestly don't know about ta.window, not my code.

Comment (happppp) [reply src/devices/cpu/mcs51/mcs51.cpp]: Good, did you check that it still works as expected or does the delay need to be increased?

Comment (DavidHaywood) [reply src/devices/cpu/mcs51/mcs51.cpp]: it appears to be fine

**Pull request merged**